Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -1355,6 +1394,26 @@ extension HTTP2FrameDecoder {
         return body(&self.state)
     }
 
+    /// This is the implementation for a throwing function. If the state doesn't get updated because an error
+    /// was thrown, this function sets the parser state to the initial one.
+    ///
+    /// The `body` function must assign a value to the parser state, otherwise it will be the intermediary
+    /// `appending` state, which is not wanted and it results in the this function's assertion failure.
+    @inline(__always)
+    private mutating func avoidingParserCoW<ReturnType>(_ body: (inout ParserState) throws -> ReturnType) throws -> ReturnType {
```

Comment:
Rather than adding a complete new function can we just modify the existing one to accept a throwing closing and use `rethrows` to propagate the throwing-ness of the closure?